### PR TITLE
[READY] Adds waddle quirk and drunk side effect

### LINKED ISSUE #8745: Add waddling quirk
[Title]: Add waddling quirk

[Implementation]: Should be easy to add the new component on quirk gain

[Why]: Would be funny to see people waddle about the place

[Resources]: None

# Be sure to ask some maintainer to label this issue as a feature. Either ping them on our discord or PM them on our forum. Thank you for sharing your ideas!
@HippieStation/maintainers

### PR DESCRIPTION
**Pull request by JohnGinnane**
the `has_quirk` proc doesn't work so I had to use a loop to check for roundstart_quirks so to not remove the component if someone with the quirk sobers up

:cl: JohnGinnane
add: Added a waddle quirk for no cost
add: Getting drunk adds the waddle effect, scaled based on how drunk you are
/:cl:

[why]: Fixes #8745

### CODE REVIEW

**Comment by ghost** [comment]
+1

File: hippiestation.dme [hunk 1]
```diff
@@ -370,6 +370,7 @@
 #include "code\datums\components\swarming.dm"
 #include "code\datums\components\thermite.dm"
 #include "code\datums\components\uplink.dm"
+#include "code\datums\components\waddling.dm"
```
Comment (ghost) [inline hippiestation.dme:373]: why include the /tg/ version if we run our own version?

File: hippiestation/code/datums/components/waddling.dm [hunk 2]
```diff
@@ -0,0 +1,15 @@
+/datum/component/waddling
```
Comment (ghost) [inline hippiestation/code/datums/components/waddling.dm:1]: `dupe_mode = COMPONENT_DUPE_UNIQUE_PASSARGS` if you choose to stop running the /tg/ version

File: hippiestation/code/datums/components/waddling.dm [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+	var/waddle_min = -12
+	var/waddle_max = 12
+	var/z_change = 4
+	var/anim_time = 2
+
```
Comment (ghost) [inline hippiestation/code/datums/components/waddling.dm:6]:
```
/datum/component/waddling/Initialize()
    if(!isliving(parent))
        return COMPONENT_INCOMPATIBLE
    RegisterSignal(parent, list(COMSIG_MOVABLE_MOVED), .proc/Waddle)
```
in the event you don't run /tg/ waddle

File: hippiestation/code/modules/mob/living/carbon/life.dm [hunk 4]
```diff
@@ -8,3 +8,40 @@
...
+		if (istype(H))
+			var/datum/component/waddling/W = H.GetComponent(/datum/component/waddling)
+
+			if (!W)
+				H.AddComponent(/datum/component/waddling)
+				to_chat(src, "<span class='warning'>Walking straight feels very hard...</span>")
+				W = H.GetComponent(/datum/component/waddling)
+
+			// minimum of 1, max of 4
+			var/waddle_multi = min(max(1, drunkenness / 25), 4)
```
Comment (ghost) [inline hippiestation/code/modules/mob/living/carbon/life.dm]: use CLAMP() here
https://github.com/HippieStation/HippieStation/blob/master/code/__DEFINES/maths.dm#L33
functionally the same but clamp is easier to read

File: hippiestation/code/modules/mob/living/carbon/life.dm [hunk 5]
```diff
@@ -8,3 +8,40 @@
...
+
+		if (!has_waddle)
+			var/mob/living/carbon/human/H = src
+
+			if (istype(H))
+				var/datum/component/waddling/W = H.GetComponent(/datum/component/waddling)
+
+				if (W)
+					W.RemoveComponent()
+					to_chat(src, "<span class='warning'>Walking doesn't seen so hard as you sober up</span>")
```
Comment (ghost) [inline hippiestation/code/modules/mob/living/carbon/life.dm]: `sober up` to `sober up.`

File: hippiestation/code/datums/traits/neutral.dm [hunk 6]
```diff
@@ -96,4 +96,23 @@
 	var/mob/living/carbon/human/H = quirk_holder
 	if(H)
 		H.scream_vol = initial(H.scream_vol)
-		H.scream_oxyloss = initial(H.scream_oxyloss)
\ No newline at end of file
+		H.scream_oxyloss = initial(H.scream_oxyloss)
+
+/datum/quirk/waddle
+	name = "Waddle"
+	desc = "You have an extra bounce in your step"
```
Comment (ghost) [inline hippiestation/code/datums/traits/neutral.dm]: `step.`

**Review by ghost — DISMISSED** [review]

File: hippiestation.dme (hunk 1 above)
Comment (JohnGinnane) [reply hippiestation.dme:373]: I guess in case they make changes to it, idk if it's a great idea to mirror the entire file, but let me know if i should

**Review by ghost — DISMISSED** [review]

File: hippiestation/code/modules/mob/living/carbon/life.dm [hunk 7]
```diff
@@ -8,3 +8,40 @@
 	for(var/T in get_traumas())
 		var/datum/brain_trauma/BT = T
 		BT.on_life()
+
+/mob/living/carbon/handle_status_effects()
```
Comment (Carbonhell) [inline hippiestation/code/modules/mob/living/carbon/life.dm:12]: if this code is intended for humans why is it at carbon level
can mobs other than humans even get drunk

File: hippiestation/code/modules/mob/living/carbon/life.dm [hunk 8]
```diff
@@ -8,3 +8,40 @@
...
+	if (drunkenness > 20)
+		var/mob/living/carbon/human/H = src
+
+		if (istype(H))
+			var/datum/component/waddling/W = H.GetComponent(/datum/component/waddling)
+
+			if (!W)
+				H.AddComponent(/datum/component/waddling)
+				to_chat(src, "<span class='warning'>Walking straight feels very hard...</span>")
+				W = H.GetComponent(/datum/component/waddling)
```
Comment (Carbonhell) [inline hippiestation/code/modules/mob/living/carbon/life.dm]: doesn't AddComponent return the newly created component?

File: hippiestation/code/modules/mob/living/carbon/life.dm [hunk 9]
```diff
@@ -8,3 +8,40 @@
...
+
+			// minimum of 1, max of 4
+			var/waddle_multi = CLAMP(drunkenness / 25, 1, 4)
+
+			W.waddle_min = -12 * waddle_multi
+			W.waddle_max = 12 * waddle_multi
+			W.z_change = 4 * waddle_multi / 2
+	else
+		var/has_waddle = FALSE
+		for(var/datum/quirk/Q in roundstart_quirks)
```
Comment (Carbonhell) [inline hippiestation/code/modules/mob/living/carbon/life.dm]: if(locate(/datum/quirk/waddle) in roundstart_quirks) ?

File: hippiestation/code/modules/mob/living/carbon/life.dm [hunk 10]
```diff
@@ -8,3 +8,40 @@
...
+
+		if (!has_waddle)
+			var/mob/living/carbon/human/H = src
+
+			if (istype(H))
+				var/datum/component/waddling/W = H.GetComponent(/datum/component/waddling)
+
+				if (W)
+					W.RemoveComponent()
+					to_chat(src, "<span class='warning'>Walking doesn't seen so hard as you sober up.</span>")
```
Comment (Carbonhell) [inline hippiestation/code/modules/mob/living/carbon/life.dm]: seem

File: hippiestation/code/datums/components/waddling.dm (hunk 2 above)
Comment (Carbonhell) [reply hippiestation/code/datums/components/waddling.dm:1]: so what is this exactly,a full mirroring or just some added vars? the latter,right?

File: hippiestation/code/modules/mob/living/carbon/life.dm (hunk 7 above)
Comment (JohnGinnane) [reply hippiestation/code/modules/mob/living/carbon/life.dm:12]: I don't decide thit homie, I just searched for where drunkenness is handled

File: hippiestation/code/modules/mob/living/carbon/life.dm (hunk 8 above)
Comment (JohnGinnane) [reply hippiestation/code/modules/mob/living/carbon/life.dm]: @McBawbaggings

File: hippiestation/code/modules/mob/living/carbon/life.dm (hunk 9 above)
Comment (JohnGinnane) [reply hippiestation/code/modules/mob/living/carbon/life.dm]: I tried, didn't work

File: hippiestation/code/modules/mob/living/carbon/life.dm (hunk 8 above)
Comment (Carbonhell) [reply hippiestation/code/modules/mob/living/carbon/life.dm]: https://github.com/HippieStation/HippieStation/blob/master/code/datums/components/_component.dm#L265

File: hippiestation/code/modules/mob/living/carbon/life.dm [hunk 11]
```diff
@@ -8,3 +8,40 @@
...
+	. = ..()
+
+	if (drunkenness > 20)
+		var/mob/living/carbon/human/H = src
+
+		if (istype(H))
+			var/datum/component/waddling/W = H.GetComponent(/datum/component/waddling)
+
+			if (!W)
+				H.AddComponent(/datum/component/waddling)
```
Comment (Carbonhell) [inline hippiestation/code/modules/mob/living/carbon/life.dm]: instead of this you could just use LoadComponent. if the component exists,it returns that, else it creates a new one. It also returns it, to avoid the next getcomponent call,i believe.

File: hippiestation/code/datums/components/waddling.dm (hunk 2 above)
Comment (ghost) [reply hippiestation/code/datums/components/waddling.dm:1]: latter, with Initialize() overrided

**Comment by HippieBartels** [comment]
+1

**Comment by Eagle-Eyess** [comment]
+1 I’d love this as a trait

**Comment by JohnGinnane** [comment]
it will be a trait, they were renamed to quirks

**Comment by Carbonhell** [comment]
is this ready?

**Comment by JohnGinnane** [comment]
not yet, I haven't address your last comments, will do later this week

File: hippiestation/code/modules/mob/living/carbon/life.dm (hunk 11 above)
Comment (JohnGinnane) [reply hippiestation/code/modules/mob/living/carbon/life.dm]: Just testing this, and that's not gonna work as I only want to display the text when the component doesn't exist and I'm adding it on this call, every subsequent call should have the component and not display the message

**Comment by JohnGinnane** [comment]
This is ready

File: hippiestation/code/datums/components/waddling.dm [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+/datum/component/waddling/Waddle()
+	var/mob/living/L = parent
```
Comment (Carbonhell) [inline hippiestation/code/datums/components/waddling.dm:8]: are you sure parent can only be a living?what if someone tries to assign it to a mob?components should have a proc that deletes them if they're given to an incompatible object, you may want to use that?

File: hippiestation/code/modules/mob/living/carbon/life.dm [hunk 13]
```diff
@@ -8,3 +8,40 @@
...
+		var/mob/living/carbon/human/H = src
+
+		if (istype(H))
+			var/datum/component/waddling/W = H.GetComponent(/datum/component/waddling)
+
+			if (!W)
+				W = H.AddComponent(/datum/component/waddling)
+				to_chat(src, "<span class='warning'>Walking straight feels very hard...</span>")
+			
+			if (W)
```
Comment (Carbonhell) [inline hippiestation/code/modules/mob/living/carbon/life.dm]: isn't this redundant? since you're creating it if it doesn't exist, it must exist here,no?

File: hippiestation/code/modules/mob/living/carbon/life.dm [hunk 14]
```diff
@@ -8,3 +8,40 @@
...
+				// minimum of 1, max of 4
+				var/waddle_multi = CLAMP(drunkenness / 25, 1, 4)
+
+				W.waddle_min = -12 * waddle_multi
+				W.waddle_max = 12 * waddle_multi
+				W.z_change = 4 * waddle_multi / 2
+	else
+		var/has_waddle = FALSE
+		for(var/datum/quirk/Q in roundstart_quirks)
```
Comment (Carbonhell) [inline hippiestation/code/modules/mob/living/carbon/life.dm]: why the type?just do for(var/q in roundstart_quirks) then var/datum/quirk/Q = q
the list can only have quirks,i assume

File: hippiestation/code/modules/mob/living/carbon/life.dm [hunk 15]
```diff
@@ -8,3 +8,40 @@
...
+			if (istype(Q, /datum/quirk/waddle))
```
Comment (Carbonhell) [inline hippiestation/code/modules/mob/living/carbon/life.dm]: wait, what is this? just do has_waddle = locate(/datum/quirk/waddle) in roundstart_quirks without the for loop and this if check, it's way faster

File: hippiestation/code/datums/components/waddling.dm (hunk 12 above)
Comment (ghost) [reply hippiestation/code/datums/components/waddling.dm:8]: yes
https://github.com/HippieStation/HippieStation/blob/master/code/datums/components/waddling.dm#L4

File: hippiestation/code/modules/mob/living/carbon/life.dm (hunk 13 above)
Comment (JohnGinnane) [reply hippiestation/code/modules/mob/living/carbon/life.dm]: it should but better safe than sorry, don't want run times

Comment (Carbonhell) [reply hippiestation/code/modules/mob/living/carbon/life.dm]: I mean, if this happens it means AddComponent fucked up badly, which would mean something is deeply fucked within core component code.
https://github.com/HippieStation/HippieStation/blob/master/code/datums/components/_component.dm
It would hinder a serious problem rather than fixing it

Comment (JohnGinnane) [reply hippiestation/code/modules/mob/living/carbon/life.dm]: 🤔 so take it out?

Comment (Carbonhell) [reply hippiestation/code/modules/mob/living/carbon/life.dm]: Honestly yeah. Maybe just do a quick check if everything runs correctly, too.

**Comment by JohnGinnane** [comment]
@Carbonhell

**Review by Carbonhell — APPROVED** [review]
